nailgun/entities.py

```diff
@@ -423,74 +422,96 @@ def __init__(self, server_config=None, **kwargs):
+        Also, do not try to read the "password" field. No value is returned for
+        the field, for obvious reasons.
+
+        """
+        if attrs is None:
+            attrs = self.read_json()
+        if 'email' not in attrs and 'email' not in ignore:
+            response = client.put(
+                self.path('self'),
+                {},
+                **self._server_config.get_client_kwargs()
+            )
+            response.raise_for_status()
+            attrs['email'] = response.json()['email']
+        return super(DockerComputeResource, self).read(entity, attrs, ignore)
 
-        Depending upon the value of ``self.provider``, various other fields are
-        filled in with values too.
-
-        .. WARNING: This method is fragile and needs significant work.
-
-        """
-        if not hasattr(self, 'provider'):
-            self.provider = self._fields['provider'].gen_value()
-
-        # Deal with docker provider before calling super create_missing in
-        # order to check if an URL is provided by the user and, if not,
-        # generate an URL pointing to a docker server
-        if self.provider.lower() == 'docker':
-            if not hasattr(self, 'url'):
-                self.url = '{0}:2375'.format(self._server_config.url)
-
-        # Now is good to call super create_missing
-        super(ComputeResource, self).create_missing()
-
-        # Generate required fields according to the provider. First check if
-        # the field is already set by the user, if not generate a random value
-        if self.provider in ('EC2', 'Ovirt', 'Openstack'):
-            for field in ('password', 'user'):
-                if not hasattr(self, field):
-                    setattr(self, field, self._fields[field].gen_value())
-        elif self.provider == 'GCE':
-            # self.email = self._fields['email'].gen_value()
-            # self.key_path = self._fields['key_path'].gen_value()
-            # self.project = self._fields['project'].gen_value()
-            #
-            # NOTE: These three pieces of data are required. However, the API
-            # docs don't even mention their existence!
-            #
-            # 1. Figure out valid values for these three fields.
-            # 2. Uncomment the above.
-            # 3. File an issue on bugzilla asking for the docs to be expanded.
-            pass
-        elif self.provider == 'Rackspace':
-            # NOTE: Foreman always returns this error:
-            #
-            #     undefined method `upcase' for nil:NilClass
-            #
-            # 1. File a bugzilla issue asking for a fix.
-            # 2. Figure out what data is necessary and add it here.
-            pass
-        elif self.provider == 'Vmware':
-            for field in ('password', 'user', 'uuid'):
-                if not hasattr(self, field):
-                    setattr(self, field, self._fields[field].gen_value())
+
+class LibvirtComputeResource(AbstractComputeResource):
```

Comment:
NailGun ignores any extra attributes that are returned. In addition to `compute_attributes` and `images`, the server also returns the `created_at` and `updated_at` attributes, and we ignore them too.